core/src/HIP/Kokkos_HIP_Instance.cpp

```diff
@@ -236,25 +236,45 @@ Kokkos::HIP::size_type *HIPInternal::scratch_flags(const std::size_t size) {
   return m_scratchFlags;
 }
 
-Kokkos::HIP::size_type *HIPInternal::scratch_functor(
-    const std::size_t size) const {
+Kokkos::HIP::size_type *HIPInternal::stage_functor_for_execution(
+    void const *driver, std::size_t const size) const {
   if (verify_is_initialized("scratch_functor") && m_scratchFunctorSize < size) {
     m_scratchFunctorSize = size;
 
     using Record = Kokkos::Impl::SharedAllocationRecord<Kokkos::HIPSpace, void>;
+    using RecordHost =
+        Kokkos::Impl::SharedAllocationRecord<Kokkos::HIPHostPinnedSpace, void>;
 
-    if (m_scratchFunctor)
+    if (m_scratchFunctor) {
       Record::decrement(Record::get_record(m_scratchFunctor));
+      RecordHost::decrement(RecordHost::get_record(m_scratchFunctorHost));
+    }
 
     Record *const r =
         Record::allocate(Kokkos::HIPSpace(), "Kokkos::InternalScratchFunctor",
                          m_scratchFunctorSize);
+    RecordHost *const r_host = RecordHost::allocate(
+        Kokkos::HIPHostPinnedSpace(), "Kokkos::InternalScratchFunctorHost",
+        m_scratchFunctorSize);
 
     Record::increment(r);
+    RecordHost::increment(r_host);
 
-    m_scratchFunctor = reinterpret_cast<size_type *>(r->data());
+    m_scratchFunctor     = reinterpret_cast<size_type *>(r->data());
+    m_scratchFunctorHost = reinterpret_cast<size_type *>(r_host->data());
   }
 
+  // When using HSA_XNACK=1, it is necessary to copy the driver to the host to
+  // ensure that the driver is not destroyed before the computation is done.
+  // Without this fix, all the atomic tests fail. It is not obvious that this
+  // problem is limited to HSA_XNACK=1 even if all the tests pass when
+  // HSA_XNACK=0. That's why we always copy the driver.
+  KOKKOS_IMPL_HIP_SAFE_CALL(hipStreamSynchronize(m_stream));
```

Comment:
I might add a note in the comment that the `hipMemcpyAsync` below should guard from anyone overwriting the device `scratchFunctor` while the current kernel is executing. My read is that this synchronization actually functions by guarding anyone from copying to the `scratchFunctorHost` while is still possibly needed for the `hipMemcpyAsync`.

I am not sure I see a better way to ensure this, but it might be useful context if we ever have to monkey with this again.